Fixed. rest api input error

Lupino opened the pull request.

dkeeney commented on src/htm/ntypes/ArrayBase.cpp:
```diff
@@ -525,31 +525,6 @@ std::istream &operator>>(std::istream &inStream, ArrayBase &a) {
-
 // Serialization and Deserialization using YAML parser
 // For JSON, expecting something like {type: "Int32", data: [1, 0, 1]}
 void ArrayBase::fromYAML(const std::string &data) { // handles both YAML and JSON
```
This function is a low level function that is used for more than REST Api.  Let's leave it as it was, including the parseDim( ) function, and create a new function with the following signature.
```
    void ArrayBase::fromVaue(const Value& vm)
```
Then in RESTapi::put_input_request( ... )
- confirm that the input has been initialized...or just call itr->second.net->initialize( )
- fetch the Input's pre-allocated buffer  (the Array object)
- parse the input string into a Value object.
- verify that the number of data elements is consistent with the input's dimensions.  ( getCount( ) ).  Throw an exception if they don't match.
- call this new fromValue( ) function on the buffer to directly set its value.

The user does not need to pass in the type or the dimensions.

dkeeney commented on src/htm/ntypes/ArrayBase.cpp:
```diff
@@ -562,88 +536,107 @@ void ArrayBase::fromYAML(const std::string &data) { // handles both YAML and JSO
+       sparse.push_back(x);
+     }
+     sdr.setSparse(sparse);
+     return;
```
Let me think about this and maybe I can suggest some logic that would work.

dkeeney commented:
Sorry, my review was no very clear.  Perhaps if I explained it using code.  
The following is a suggestion which would pass the data (without needing type or dimensions).

In RESTapi.cpp
```
std::string RESTapi::put_input_request(const std::string &id,
                                       const std::string &region_name,
                                       const std::string &input_name,
                                       const std::string &data) {
  try {
    auto itr = resource_.find(id);
    NTA_CHECK(itr != resource_.end()) << "Context for resource '" + id + "' not found.";
    itr->second.t = time(0);

    // Expected message is JSON, {"data": [ 1, 1, 0, 1]}
    Value vm;
    vm.parse(data);
    NTA_CHECK(vm.contains("data"))
        << "Unexpected YAML or JSON format. Expecting something like {\"data\": [1,2,3]}";
    Value vm1 = vm["data"];
    NTA_CHECK(vm1 && vm1.isSequence())
      << "Unexpected YAML or JSON format. Expecting something like {\"data\": [1,2,3]}";
    
     // make sure buffers have been created.
    itr->second.net->initialize();   
    
    // get a reference to the allocated buffer for the input.
    itr->second.net->getRegion(region_name)->getInputData(input_name); 
    
    // Copy the data into it with type conversions.
    // Type and dimensions are not changed.
    // Throws an exception if number of elements does not match elements in allocated dimensions.
    a.fromValue(vm1);

    return "{\"result\": \"OK\"}";
  }
  catch (Exception &e) {
    return "{\"err\": " + Value::json_string(e.getMessage()) + "}";
  } catch (std::exception& e) {
    return "{\"err\": " + Value::json_string(e.what()) + "}";
  } catch (...) {
    return "{\"err\": " + Value::json_string("Unknown Exception.") + "}";
  }
}
```

in Region.cpp
You no longer call setInputData( ) but please fix this for C++ apps that may want to set the input from data in an Array object.
```
void Region::setInputData(const std::string &inputName, const Array& data) {
  network_->initialize();
  Array& a = getInputData(inputName);
  NTA_CHECK(a.getCount() == data.getCount()) << "Unexpected number elements. Expected " << a.getCount() << " found " << data.getCount() << ".";
  data.convertInto(a);  // Copies buffer in data into buffer in a, performs type conversion if needed.
}
```


In ArrayBase.cpp
Lets break up fromYAML( ) so the bottom part can be fromValue( ).
You will not need fromYAML( ) but others will.  The syntax matches that which is output with the << operator.
Note that the parseDim( ) function is still needed.
```
// Serialization and Deserialization using YAML parser
// For JSON, expecting something like {type: "Int32", data: [1, 0, 1]}
void ArrayBase::fromYAML(const std::string &data) { // handles both YAML and JSON
  size_t num;
  Value vm, vm1, vm2;
  vm.parse(data);
  NTA_CHECK(vm.contains("type") && vm.contains("data"))
      << "Unexpected YAML or JSON format. Expecting something like {type: \"Real32\", data: [1,0,1]}";

  vm1 = vm["type"];
  NTA_CHECK(vm1 && vm1.isScalar())
      << "Unexpected YAML or JSON format. Expecting something like {type: \"Int32\", data: [1,0,1]}";
  vm2 = vm["data"];
  NTA_CHECK(vm2 && vm.isSequence())
      << "Unexpected YAML or JSON format. Expecting something like {type: \"SDR(1000)\", data: [1,2,3]}";

  std::string typeStr = vm1.as<std::string>();
  if (typeStr.size() >= 3 && typeStr.substr(0, 3) == "SDR") {
    type_ = NTA_BasicType_SDR;
    allocateBuffer(parseDim(typeStr));
  } else {
    type_ = BasicType::parse(typeStr);
    size_t num = vm2.size();
    allocateBuffer(num);
  }
  fromValue(vm2);
}

// This will populate this Array with the parsed data.
void ArrayBase::fromValue(const Value& vm)
  num = vm2.size();
  void *inbuf = getBuffer();
  
  NTA_CHECK(num == getCount()) << "Parsed data has " << num << " elements.  Expected " << getCount() << " elements.\n";

  switch (type_) {
  case NTA_BasicType_Byte:
    for (size_t i = 0; i < num; i++) {
      ((Byte *)inbuf)[i] = vm2[i].as<Byte>();
    }
    break;
  case NTA_BasicType_Int16:
    for (size_t i = 0; i < num; i++) {
      ((Int16 *)inbuf)[i] = vm2[i].as<Int16>();
    }
    break;
  case NTA_BasicType_UInt16:
    for (size_t i = 0; i < num; i++) {
      ((UInt16 *)inbuf)[i] = vm2[i].as<UInt16>();
    }
    break;
  case NTA_BasicType_Int32:
    for (size_t i = 0; i < num; i++) {
      ((Int32 *)inbuf)[i] = vm2[i].as<Int32>();
    }
    break;
  case NTA_BasicType_UInt32:
    for (size_t i = 0; i < num; i++) {
      ((UInt32 *)inbuf)[i] = vm2[i].as<UInt32>();
    }
    break;
  case NTA_BasicType_Int64:
    for (size_t i = 0; i < num; i++) {
      ((Int64 *)inbuf)[i] = vm2[i].as<Int64>();
    }
    break;
  case NTA_BasicType_UInt64:
    for (size_t i = 0; i < num; i++) {
      ((UInt64 *)inbuf)[i] = vm2[i].as<UInt64>();
    }
    break;
  case NTA_BasicType_Real32:
    for (size_t i = 0; i < num; i++) {
      ((Real32 *)inbuf)[i] = vm2[i].as<Real32>();
    }
    break;
  case NTA_BasicType_Real64:
    for (size_t i = 0; i < num; i++) {
      ((Real64 *)inbuf)[i] = vm2[i].as<Real64>();
    }
    break;
  case NTA_BasicType_Bool:
    for (size_t i = 0; i < num; i++) {
      ((bool *)inbuf)[i] = vm2[i].as<bool>();
    }
    break;
  case NTA_BasicType_SDR: //  Expecting dense data
  {
    SDR &sdr = getSDR();
    SDR_dence_t dence;
    for (size_t i = 0; i < num; i++) {
      UInt x = vm2[i].as<Byte>();
      dence.push_back(x);
    }
    sdr.setDense(dence);
    break;
  }
  case NTA_BasicType_Str: 
    for (size_t i = 0; i < num; i++) {
      ((std::string *)inbuf)[i] = vm2[i].as<std::string>();
    }
    break;
  default:
    NTA_THROW << "Unexpected Element Type: " << type_;
    break;
  }
}
```

And note that this was fixed to expect dense format for SDR rather than sparse format.  This would break the >> operator if we used it with data provided in the << operator.   Not sure what to do about that.

dkeeney commented:
@Lupino, I don't want you to get discouraged with all of the requested changes.  Much of this is code that I did not think out entirely when I wrote it.

The code I presented above will probably not compile and I probably am missing some important thing but I think it gives you an idea of what I was thinking.

Lupino commented:
in the real situation the Input not know its dim.

when i pass a SDR input to SPRegion in the eg. 

the sp bottomUpIn do not know its size,  an there is not out links,

when  SPRegion.initialize, must make sure inputs initialized,  in the follow eg. the sp inputs is not initialized

so Region.setInputData before Network.initialized is not an good ideal. 

```
  def testNetworkRESTSetInputSdr(self):

    net = NetworkREST(host=HOST, verbose=True)

    sp = net.add_region('sp', 'SPRegion', {
      'columnCount': 2048,
      'globalInhibition': True
    })

    net.create()

    r = sp.input('bottomUpIn', [1,2], [100])
    self.assertEqual(r, 'OK')

    # Execute one iteration of the Network object
    r = net.run(1)
    self.assertEqual(r, 'OK')
```

Lupino commented on src/htm/engine/Region.cpp:
```diff
@@ -446,15 +446,16 @@ const Array& Region::getInputData(const std::string &inputName) const {
   std::shared_ptr<Input> in = ii->second;
-	in->setDimensions( { (UInt)data.getCount() } );
+  in->initialize();
+  in->setDimensions( { (UInt)data.getCount() } );
```
the Dimensions is wrong, if there is no any links.

Lupino commented on src/htm/engine/RESTapi.cpp:
```diff
@@ -86,19 +87,24 @@ std::string RESTapi::create_network_request(const std::string &specified_id, con
+    Array a(destSpec->inputs.getByName(input_name).dataType);
 
-    itr->second.net->getRegion(region_name)->setInputData(input_name, a);
+    a.fromValue(vm);
```
the real usage,  the type is right, but dimensions maybe wrong,          when the input has not links.

Lupino commented:
SDR use  setDense is difference of SDR << operator

-- The syntax matches that which is output with the << operator.

dkeeney commented:
> in the real situation the Input not know its dim.

> so Region.setInputData before Network.initialized is not an good ideal.

I think it depends on how we expect this to be used.
case 1:  Before initialization.  The client sets the input only once.  (perhaps to set an initial value).
case 2:  After initialization. The client sets the input repeatedly, once for each run in a loop. (To feed iterative data to the Region)

For case1, you are right, the dimensions are not known by the input and we can pass the dimensions in the message and use that to set the input buffer size.  Then during initialization, the Link logic will propagate that dimension to the connected regions.

For case2, the dimensions and buffer sizes have already been set during initialization and cannot be changed.  The SP and TM algorithms assume that SDR dimensions are the same for each iteration.   For the encoders, we might be able to change data width but the design assumption was that the buffer widths are fixed.

I was thinking case 2.  That the user wants to repeatedly use this to feed data to a region's input between runs.

Do we want to allow both cases?   I am not opposed to that as long as the dimensions do not change after initialization.
I know this is confusing.  Perhaps that is why the original Numenta's NuPIC code for NetworkAPI did not allow putting data directly into the input buffers.

Lupino commented:
In the ClassifierRegion's bucket input,    maybe difference size on every time puts.

egs.

```
SDR_a   -----   cls_1, cls_2
SDR_b   -----   cls_1,cls_2,cls_3
```

so the input size must change every time on the real puts data

dkeeney commented:
> In the ClassifierRegion's bucket input,

I do not understand your example.  Perhaps you can explain it a little more.

The ClassifierRegion (and the SDRClassifier algorithom) is expecting two inputs
- A bucket which is an index to a category.  This is a scalar.  Only one value. Normally the bucket output of the encoder which is a quantized version of its input.  
- An pattern which is an SDR  (this is fixed size).  This is normally the output of the SP or TM which was a result of the bucket being encoded.  The dimensions of the SDR are fixed during initialization.

The algorithm will train on the bucket and pattern pairs such that eventually when it sees a particular pattern it can match it up with its corresponding bucket.  The output is then an array of probabilities of being a match of a bucket for this pattern.

So I do not see where the input (or output) dimensions need to change after initialization.

The probability (pdf) output of the ClassifierRegion is the only case I can think of which can change in size and this is not an SDR.  The width of the output array is the number of unique bucket values it has seen while training.  As it sees more new bucket values it will grow the array.  That output cannot be fed into any other input...it would break.

Lupino commented:
This is an example to make a tag of a movie base it's name.

one movie has one tag or more, the follow is my code.

the bucket  size is changes, for every movie. 


```python
import os
import re
import random
from htm_rest_api import NetworkREST

# http://117.128.6.34/cache/files.grouplens.org/datasets/movielens/ml-100k.zip?ich_args2=528-05191115033744_9b125d263d0c20015f9dd28d5a04fa11_10001002_9c896d2fdec7f6d4973e518939a83798_cb8cbd8c00b1c8d60975f48b4864adb0

dataset = 'ml-100k'
re_skip = re.compile('[.(),\'-]')
skip_len = 2


# 1681|You So Crazy (1994)|01-Jan-1994||http://us.imdb.com/M/title-exact?You%20So%20Crazy%20(1994)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0|0|0|0
def parse_item(line):
  line = line.split('|')
  title = line[1]
  cats = [i for i, v in enumerate(line[5:]) if v == '1']
  return {'title': title, 'cats': cats}


def load_items():
  items = []
  with open(os.path.join(dataset, 'u.item'), 'rb') as f:
    for line in f:
      line = str(line, 'utf-8', errors='ignore').strip()
      if line:
        items.append(parse_item(line))

  return items


def title2words(title):
  title = re_skip.sub(' ', title).lower()
  return [w for w in title.split(' ') if len(w) > skip_len]


def load_words(items):
  words = []
  for item in items:
    for w in title2words(item['title']):
      if w not in words:
        words.append(w)

  return words


def title2sparse(words, title):
  sparse = []
  for w in title2words(title):
    sparse.append(words.index(w))

  return sparse


def get_predict(clsr, min_score=0.1):
  cats = clsr.output('titles')
  pdf = clsr.output('pdf')

  ret = []

  for cat, p in zip(cats, pdf):
    if p > min_score:
      ret.append((cat, p))

  return ret


# line = '1681|You So Crazy (1994)|01-Jan-1994||http://us.imdb.com/M/title-exact?You%20So%20Crazy%20(1994)|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0|0|0|0'
# print(parse_item(line))

items = load_items()

words = load_words(items)

dim = [len(words)]

random.shuffle(items)

train_size = int(len(items) * 0.8)

train_items = items[:train_size]
test_items = items[:train_size]

# for item in items:
#     print(title2sparse(words, item['title']))

# Setup the AI.
net = NetworkREST(verbose=False)
sp = net.add_region(
  'sp',
  'SPRegion',
  dict(
    columnCount=5000,
    globalInhibition=True,
    seed=
    0,  # this is important, 0="random" seed which changes on each invocation
    spVerbosity=99,
    wrapAround=False))

clsr = net.add_region('clsr', 'ClassifierRegion', {'learn': True})
net.add_link(sp, clsr, 'bottomUpOut', 'pattern')

net.create()

# Training Loop
for i, item in enumerate(train_items):
  sp.input('bottomUpIn', title2sparse(words, item['title']), dim)
  clsr.input('bucket', item['cats'])

  net.run(50)

  if i % 100 == 0:
    print('iter: ', i, '/', train_size)
    pred = get_predict(clsr)
    print(pred, item['cats'])

# Testing Loop
score = 0
total = 0
sp.param('learningMode', 0)
clsr.param('learn', 'false')
for item in test_items:
  sp.input('bottomUpIn', title2sparse(words, item['title']), dim)
  net.run()
  pred = get_predict(clsr)
  for r, _ in pred:
    if r in item['cats']:
      score += 1

  total += len(item['cats'])
  print(pred, item['cats'])

score = score / total

print('Score:', 100 * score, '%')
```

dkeeney commented:
Thank you for your example.  Very interesting.

**variable sized buffer**
Ok, you have convinced me that a variable sized buffer (meaning that each iteration could pass a different sized array) would be useful as the input to an encoder since that data could be anything. Sense there is one case where a variable width buffer should be allowed in the REST Api protocol.

**dim field**
In your example you only show a single dimension  `dim = [len(words)]`.    The length of the dense array decoded by JSON is the same as its dimension in this case.   JSON can also encode multiple dimension arrays using [ ]'s around each row, etc.  I don't think our core code handles that when we turn that into an Array object.  That is a topic for another bug report, but this does not affect your example. Your example has only one dimension.

You show building your SDR using sparse data and I think we decided with a previous PR that, even for an SDR, the data must be in **dense** format. So I see why you needed the dim field in order to pass sparse data.  But with dense format the dim does not need to be specified.  It's treated as an array of '1's and '0's and the number of elements IS the dimension.

**Other observations**
In your example it appears that your code is acting as the encoder.  This is good.
For the semantic meaning to exist in the data, each bit must have some meaning.  In your case, for each movie the same bit should have the same meaning.  So if bit[0] was set on one movie and it was also set on another movie then those two movies have something in common.  An encoder may use one bit for a feature or it may use several bits but they will not overlap with other features unless the features are related.

The HTM School videos explain this as well as all of the rules for an encoder.

If each bit has a meaning, then there will be a maximum number of these bits in this overall bit map and it will be fixed for all data in the dataset.  That is the dimension SP is expecting and it is expecting it to be a fixed size.  Consider: it is the overlap of a bit between two SDR's that means there is at least one feature of the two SDR patterns that are in common.

So, it appears to me that you have fixed width data in your example.

**Conclusions**
We want to allow variable width arrays to be passed in the REST api.  Changing the REST api is easy.
However, the design of the NetworkAPI code assumes fixed sized buffers allocated at initialization time.  All of the current algorithms, including SP and TM currently require fixed width buffers (that is, the buffer width is fixed during initialization).  So, to use this variable width buffer we would have to make a special case where they can ONLY be used to feed data to an encoder that declares it will accept it.  

So far I have not seen an encoder that needs this capability although it could happen.
I suggest that we wait until we see one that does need this before modifying the Link logic to allow it.

Lupino commented:
It maybe too large when parsing **SDR.dense** to REST Server every time.

Only **ClassifierRegion.bucket** accept difference size.

dkeeney commented:
> It maybe too large when parsing SDR.dense to REST Server every time.

Using dense format for SDR is not very efficient I agree.  It is not too late to change that.   Things that we must consider if we do this:
- Both client and server must know that if the type is SDR then the format will be sparse.  We can document that and make that the standard.
- Both client and server must know the type.  Do we assume that the programmer on the client side will always know the type that is expected in the connected Region or do we require the client to pass the type (and dim if an SDR) so the server side Region knows what the client sent.  I think that is what you would like to do.
- If the type is not an SDR, do we require the client to send the 'dim' or can we use the buffer size as the 'dim'?  
- In either case if the 'dim' is other than the expected, it will throw an error.
- The JSON decoder/encoder in the ArrayBase class will need to change and data conversion will need fixing.  Don't know what all that might affect.

> Only ClassifierRegion.bucket accept difference size.

No, ClassifierRegion.bucket input accepts only a scalar.  Size is 1 and is fixed.  It is the numeric index for the category that was encoded.

Lupino commented:
it is an easy way to detect a SDR input is  dense or  sparse.

I think no need a type for SDR

```python
def detect_dense(data):
    data_len = len(data)
    if data_len == 0:
        raise Exception('Empty input')

    if data[0] > 1:
        return False

    if data_len == 1:
        raise Exception('Unknow type')

    if data[1] > 1:
        return False

    if data_len == 2:
        if data[0] == data[1]:
            return True

        else:
            raise Exception('Unknow type')

    if data[2] > 1:
        return False

    if data[0] == data[2] or data[1] == data[2]:
        return True

    raise Exception('Unknow type')
```

dkeeney commented:
Sparse data stored in the SDR is a vector of indexes of elements that are non-zero. 
If we use that definition then a sparse array with zero length is valid.  It would be an SDR with all values at 0.
A length of 1 would also be valid, meaning at least one bit is on.

In fact a sparse array would be indistinguishable from a dense array of integers except that the sparse array would contain all positive values, not contain duplicate values and each successive value would be larger than the previous.  But it is possible to have a dense array of type UInt32 that happens to have such values.

Since we do type conversions we must be ready to assume that someone might want to send us a dense array of any type and expect it to be converted into an SDR.  If the SDR format used dense format then it does not matter.  Each element is 0 or non-zero and conversion can produce the expected array.  But if the SDR format is sparse then I still think we need something else to tell us the type and dimensions.

From your sample code it appears that you are encoding the sparse array a little differently or making different assumptions.  I have seen some ways to serialize a sparse two dimensional matrix which include lengths or offsets.  Perhaps that is what you had in mind.  Does numpy provide a serialized version of a sparse array?

dkeeney commented:
I am trying to think through your idea of detecting if an SDR is sparse or dense from the bit patterns.

If the target input is defined as something other than SDR type we assume dense since they are just numbers.

If the target input is defined as an SDR we could make the assumptions:
- if the number of elements equals the SDR's buffer size there is a good chance that it is in dense format.  The only sparse format case with that many values must be all bits true so in this special case, we can test if buffer size >=3 and if data[0] == 0, data[1] == 1, and data[3] == 3 which would confirm that it is sparse.  Otherwise it is dense.  If it is dense it should be an error if any value is not a 0 or 'false' or 1, 1.0, or 'true'. 
- if the number of elements is **not** the same as the SDR's buffer we can assume it is sparse.  However, if a value is a negative integer or a fractional value it should be an error.  If values are not progressively increasing it should be an error.  If any value is greater than or equal to the field width it is an error.

We would have a problem if the target SDR's buffer width (it's dimension) was 1 or 2.  This could happen if this is an encoder's output of a binary value so we cannot say this is an error.  Using the above rules;
Width of 1; A value of binary 1 encoded as sparse would be [0] and interpreted as dense resulting in binary 0.
Width of 2; A value of binary 1,1 encoded as sparse would be [0,1] and interpreted as dense resulting in binary 0, 1.

These are the only two cases that I can think of where the algorithm breaks.  We can add a rule for the client side that says "if the expected SDR's dimension is 1 or 2, always encode as dense".  

Another client side rule is that if the data is to be empty, return the JSON value null without quotes or [ ].  We need to verify that this works.

dkeeney commented on src/htm/regions/SPRegion.cpp:
```diff
@@ -98,10 +98,10 @@ void SPRegion::initialize() {
   std::shared_ptr<Input> in = getInput("bottomUpIn");
   NTA_CHECK(in != nullptr);
-  if (!in->hasIncomingLinks())
+  if (!in->hasIncomingLinks() && !in->isInitialized())
```
The intention of this check was to insure that the SP had some data to work with.
Since data can now be directly assigned to an input, we can change this to something like:

NTA_CHECK(in->getDimensions().isSpecified()) << SPRegion::initialize - No links or dimensions were set for this SP region.\n";

dkeeney replied:
Lets remove this change since we will now have a link if there is data.

dkeeney commented:
The problem I am trying to solve is that we need a way to configure the dimensions on an input at initialization time when the intent is to directly feed data to an input at run time using a put input message.

I have two possible solutions.
1)  expand the 'dim' parameter to set dimensions directly on either an input or output rather than just the default dimension for the Region.  In our case it would force the input to allocate an input buffer of the appropriate size and type.

This first solution is perhaps simpler but it may not be obvious that in order to use the "push input" at runtime the 'dim' parameter must be used to set the dimensions on the target input.  Some examples of the dim syntax would be something like:
```
   dim: [1,2]                 means set the default dimension for the region.  <== current syntax
   dim: 25                    means set the default dimension to this single dimension.
   dim: {"bottomUpIn": [5]}   means set the dimensions on the input or output for region with the name "bottomUpIn"
   dim: {"bottomUpIn": [5], "bottomUpOut": [2,1000]}   means set both of these dimensions.
```
This would not be able to be used with link functions like Fan-In or the delay buffer.

2) expand the Link command to declare a dummy source with region name "EXTERNAL" and input name is a name of your choosing like "input1" or something like that, which describes your intent to set that input with your "push input" call and treat it as if it came from an output.  Another argument in the Link definition could specify the dimensions to expect.
```
  addLink: {"src": "EXTERNAL.input1", "dest": "SP.bottomUpIn", "delay": 10, "dim": [100,200]}
```
The put input command would specify  "EXTERNAL.input1" as the target.  The "EXTERNAL" is required but the "input1" is of your choosing so you can specify among several inputs you might be setting.

This second solution is more complicated but its use is more descriptive.  There is a Link descriptor describing the fact that you intend to provide input data directly to an input and you can explicitly specify its dimension.  It has the additional advantage of being able to use type translation, Fan-In and the Delay buffer features of the link. 

Both solutions would solve data flow problems in NetworkAPI apps written in C++ and Python as well as the REST api.

At this moment I am inclined to implement both solutions although that might be over kill.  I am setting up a new PR to do this.

@Lupino, @breznak, what are your thoughts?

Lupino commented on py/htm/examples/rest/hotgym.py:
```diff
@@ -135,6 +134,7 @@ def main(parameters=default_parameters, argv=None, verbose=True):
     # Call the encoders to create bit representations for each value.  These are SDR objects.
     dateRegion.param('sensedTime', int(dateString.timestamp()))
     scalarRegion.param('sensedValue', consumption)
+    clsrRegion.input('bucket', consumption)
```
in link the perdict result is difference of the examples/hotgym.py

just put the raw data to bucket，  it is the right way.

Lupino commented on src/htm/ntypes/ArrayBase.cpp:
```diff
@@ -652,12 +656,34 @@ void ArrayBase::fromValue(const Value &vm_) {
+      } else if (num == 2) { // maybe problem here
+        isDense = true;
+      } else {
+        if (vm[0] == vm[1] || vm[0] == vm[2] || vm[1] == vm[2]) {
```
the array of input maybe has an order

eg.          [3,1,2,4,0]

dkeeney replied:
The requirement for a sparse array is that it must be sorted.

breznak commented:
> The problem I am trying to solve is that we need a way to configure the dimensions on an input at initialization time when the intent is to directly feed data to an input at run time using a put input message. 

possibly I'm misunderstanding, but on high level, I think we should behave "like tensor-flow", at compile time create the graph (dimensions, links, ...) and at runtime it's fixed and only data flows. 

> expand the 'dim' parameter to set dimensions directly on either an input or output rather than just the default dimension for the Region 

this sounds reasonable to me, easier to implement, understandable. From your code example, I like the last 2 usage patterns. 

> This would not be able to be used with link functions like Fan-In or the delay buffer. 

do we use it (fan-in/out, delays)? Does it work at all? Do we want to support that in the future? 
(bit OT: see ROS2 if we wanted to have industrial applications (and networking), or consider some larger architectural change, and instead of defining "our" Regions, make them behave like PyTorch, TF layers/nodes. So we could use some existing tools from those big ecosystems)

> addLink: {"src": "EXTERNAL.input1", "dest": "SP.bottomUpIn", "delay": 10, "dim": [100,200]}
 
this looks also usable. 
I'd just decide and stick with only one of the approaches you suggested.  

> Both solutions would solve data flow problems in NetworkAPI apps written in C++ and Python as well as the REST api. 

:+1:

breznak commented:
You've had a long and deep discussion which stemmed new ideas and goals. But with respect to the original problem this PR addressed, is this ready?

dkeeney commented:
> I think we should behave "like tensor-flow", at compile time create the graph (dimensions, links, ...) and at runtime it's fixed and only data flows.

That is my thinking as well.  Even the algorithms at the lowest level require dimensions to be fixed at the point of construction. 

> do we use it (fan-in/out, delays)? Does it work at all? Do we want to support that in the future?

Yes, Yes, and I think so because they solve some basic problems.  For example, two encoder outputs can be combined into one SDR to form a composite before being passed to the SP.  This is basically what is happening within the DateEncoder.  This combining is the Fan-in feature.  Just connect two links  to the same input on the SP.

> I'd just decide and stick with only one of the approaches you suggested.

Yes, what I would like to do is implement the link approach even though it is more involved because it keeps the Fan-In.  I am working on it now on a separate PR.  When this is complete and merged, then this PR can proceed because it will then have a means to safely set the input dimensions at configure time.

breznak commented:
>  am working on it now on a separate PR. When this is complete and merged, then this PR can proceed ... 

thank you for explaining the situation, so this PR is on hold until your PR fixes the functionality. 

> For example, two encoder outputs can be combined into one SDR to form a composite before being passed to the SP. This is basically what is happening within the DateEncoder. 

I know this was just an example, what I slightly dislike about this part of NAPI is that it's reinventing the wheel (linking, delay, ...). Is there some suitable framework which we could use for that functionality? (like we did with Serialization, REST, ...) 
If not, or too complicated, I understand your position and :+1: for the Link approach.

dkeeney commented:
> so this PR is on hold until your PR fixes the functionality.

Yes.

>  what I slightly dislike about this part of NAPI is that it's reinventing the wheel (linking, delay, ...).

reinventing?  I am curious about your ideas.
Of course I did not invent NAPI.  The original design did not allow directly setting the inputs at all and I am trying to find a way to fit this in without disrupting the existing features, specifically the Fan-In and Delay.

It turns out there are not too many changes to this but its hard to figure out where to put them in so it all flows.
Let me push what I have of the new PR so we can move this part of the discussion there.

Lupino commented:
>  specifically the Fan-In and Delay.
That's good

dkeeney commented:
@Lupino you now a facility that will work when you set the buffer.  See PR#861 and checkout docs/NetworkAPI_Links.md

Many of the recommendations I gave before are no longer valid so I recommend taking a fresh look at the problem.

We still need to flush out RESTapi::put_input_request( ... ) but we can now get the buffer that has already been allocated and parse the JSON directly into it.  Something like:
```
  std::shared_ptr<Output> src = itr->second.net.getRegion("INPUT")->getOutput(source_name);  
  std::vector<Int32> dim = src->getDimensions().asVector();   // dimensions for the output (may not be needed)

  Array &buf = src->getData();                                                  // Reference to the pre-allocated buffer
  NTA_BasicType type = buf->getType( );                                  // Type on the buffer
  UInt32 count = src->getCount();                                            // number of elements in buffer

  buf.fromJSON(data);  ????
```
- ArrayBase::fromJSON( ) just calls ArrayBase::fromYAML so we are really referring to ArrayBase::fromYAML( ).

- It was intended that ArrayBase::toJSON( ) and ArrayBase::fromJSON( ) be the reverse of each other, as a way to serialize an Array object in a general case.  At the moment they don't quite do that.  To be a complete serialization, the JSON should carry the type and dimensions.

- The ArrayBase::toJSON( ) method needs to be modified to send sparse for an SDR and ArrayBase::fromYAML( ) needs to be modified to detect if it is sparse or dense in the case of an SDR (using your sparse detection algorithm).  I think this is the decision we have already determined from previous discussion.

- But the JSON serialization mentioned is slightly different than the JSON we will be getting from the REST interface `RESTapi::put_input_request( ... )` which does not carry type/dim.  In this case we want it to use the type and dim that are already on the buffer and error if the JSON data is not compatible with it.  We also must insure that processing the REST interface does not modify the type or dimensions on the existing buffer. It should just do what checks it can and populate the buffer.

You can either make a new function or see if you can find a way to make fromYAML( ) handle both.

dkeeney commented on src/htm/engine/Network.cpp:
```diff
@@ -461,7 +461,23 @@ void Network::setInputData(const std::string &sourceName, const Value& vm) {
+
+  a.fromValue(vm);
+
   // Unfinished; topic of PR #585
```
These comments can be removed now.

dkeeney commented on src/htm/engine/Network.cpp:
```diff
@@ -461,7 +461,23 @@ void Network::setInputData(const std::string &sourceName, const Value& vm) {
   Array &a =  region->getOutput(sourceName)->getData(); // populate this output buffer that will be moved to the input.
   NTA_BasicType type = a.getType();
   UNUSED(type);
```
You used 'type' so this line can come out.

dkeeney commented on py/htm/examples/rest/mnist.py:
```diff
@@ -0,0 +1,168 @@
+# ------------------------------------------------------------------------------
+# HTM Community Edition of NuPIC
+# Copyright (C) 2018-2019, David McDougall
```
@Lupino I think you should add your name here too.

dkeeney commented on src/htm/ntypes/ArrayBase.cpp:
```diff
@@ -576,74 +576,120 @@ void ArrayBase::fromYAML(const std::string &data) { // handles both YAML and JSO
+      ((std::string *)inbuf)[i] = vm[i].as<std::string>();
     }
     break;
   default:
```
Also, down in line 706 we should go back to generating sparse in the toJSON( ) function.

dkeeney commented:
>    def add_link(self, source_region, dest_region, source_output, dest_input, dim = None, delay = None):
      if not self.has_region(source_region.name):
        raise NetworkRESTError('Region {} is not found.'.format(
>         source_region.name))
E       htm_rest_api.NetworkRESTError: Region INPUT is not found.

I think what is needed here is to skip this check if the region name is the special name "INPUT".  Processing the link message in the C++ side will implicitly create the equivalent of the "INPUT" region.
